src/psi/j/executors/saga.py

```diff
@@ -0,0 +1,222 @@
+
+        The hooks are used to close the SAGA service used by this executor.
+        """
+        atexit.unregister(self._close)
+        self._close()
+
+    def _state_cb(self, saga_job: rs.job.Job, metric: object, ctx: object) -> bool:
+        jpsi_uid = saga_job._jex_uid
+        jpsi_job = self._get_jpsi_job(jpsi_uid)
+        ec = None
+
+        if saga_job.state in [rs.FAILED, rs.DONE]:
+            ec = saga_job.exit_code
+
+        logger.debug('%s --> %s', jpsi_uid, saga_job.state)
+
+        job_status = JobStatus(_STATE_MAP[saga_job.state], time.time(), exit_code=ec)
+        jpsi_job._set_status(job_status, self)
+        if self._cb:
+            self._cb.job_status_changed(jpsi_job, job_status)
+
+        return True
+
+    def submit(self, job: Job) -> None:
+        """
+        Submits the specified :class:`~psi.j.Job` to be run using SAGA.
+
+        Parameters
+        ----------
+        job
+            The :class:`~psi.j.Job` to be submitted.
+        """
+        # derive SAGA job description and submit it
+        job_mapping = _JobMapping(job)
+        with self._lock:
+            self._jobs[job.id] = job_mapping
+        job._executor = self
+
+        jd = self._job_2_descr(job)
+        saga_job = self._js().create_job(jd)
+        saga_job._jex_uid = job.id
+
+        job_mapping.saga_job = saga_job
+        saga_job.add_callback(rs.STATE, self._state_cb)
+
+        # TODO: the backend attribute is not standard; why do we need it?
+        self._update_job_status(job, JobStatus(JobState.QUEUED, metadata={'backend': self.url}))
+
+        self._update_job_status(job, JobStatus(JobState.ACTIVE))
+
+        saga_job.run()
+
+    def _get_saga_job(self, uid: str) -> rs.job.Job:
+        with self._lock:
+            return self._jobs[uid].saga_job
+
+    def _get_jpsi_job(self, uid: str) -> Job:
+        with self._lock:
+            return self._jobs[uid].jpsi_job
+
+    def list(self) -> List[str]:
+        """
+        Retrieves a list of native IDs corresponding to jobs known to this executor.
+
+        This method does not query the SAGA backend or any of the backends behind SAGA for jobs
+        not submitted through this executor. Furthermore, it simply returns the PSI-J job IDs of
+        the jobs rather than native IDs.
+        Returns
+        -------
+        A list of job IDs.
+        """
+        with self._lock:
+            return list(self._jobs.keys())
+
+    def attach(self, job: Job, native_id: str) -> None:
+        """
+        Attaches a job to a backend job.
+
+        This method does not quite do what it is meant to do at this time.
+
+        Parameters
+        ----------
+        job
+            The job to attach.
+        native_id
+            A PSI-J job ID
+
+        """
+        # This probably needs to be re-worked. I would assume that the meaning of attach()
+        # here is not to attach to a saga job, but to attach to a scheduler job
```

Comment:
I'm afraid I don't understand that comment.  What would a `scheduler job` be in this context?  Like a Slurm job?  How would one attach to that w/o going via SAGA (and thus the SAGA job)?

Reply:
It doesn't look like one can attach to a job that was submitted in another process, since it relies on `_jobs`, which is only updated by `submit()`.